sources/nvidia-migmanager/src/main.rs

```diff
@@ -0,0 +1,688 @@
+        match self {
+            NvidiaH200_141gbMigProfile::Mig7g141gb => "7g.141gb",
+            NvidiaH200_141gbMigProfile::Mig3g71gb => "3g.71gb,3g.71gb",
+            NvidiaH200_141gbMigProfile::Mig2g35gb => "2g.35gb,2g.35gb,2g.35gb",
+            NvidiaH200_141gbMigProfile::Mig1g35gb => "1g.35gb,1g.35gb,1g.35gb,1g.35gb",
+            NvidiaH200_141gbMigProfile::Mig1g18gb => {
+                "1g.18gb,1g.18gb,1g.18gb,1g.18gb,1g.18gb,1g.18gb,1g.18gb"
+            }
+        }
+    }
+}
+
+trait MigGpuProfile: for<'de> Deserialize<'de> {
+    fn get_mig_profile(&self) -> &str;
+}
+
+/// Read the config file to get MIG settings
+fn get_mig_settings<P>(config_path: P) -> Result<NvidiaMigConfig>
+where
+    P: AsRef<Path>,
+{
+    let config_str = fs::read_to_string(config_path.as_ref()).context(error::ReadConfigSnafu {
+        config_path: config_path.as_ref(),
+    })?;
+
+    let config: NvidiaMigConfig =
+        toml::from_str(&config_str).context(error::TomlDeserializationSnafu {
+            config_path: config_path.as_ref(),
+        })?;
+
+    Ok(config)
+}
+
+fn reboot_if_required() -> Result<()> {
+    let reboot_required = Path::new(REBOOT_REQUIRED_MARKER_FILE).exists();
+
+    if reboot_required {
+        info!("GPU reset is required to apply MIG Settings. Initiating reboot...");
+        command(SYSTEMCTL_PATH, ["reboot"])?;
+    } else {
+        info!("GPU reset not required.");
+    }
+
+    Ok(())
+}
+
+fn is_mig_active(mig_settings: NvidiaMigConfig, gpu_info: Vec<MigGpu>) -> Result<()> {
+    if mig_settings.device_partitioning_strategy == NvidiaDevicePartitioningStrategy::MIG {
+        for gpu in &gpu_info {
+            if gpu.model == NvidiaGpu::A100_40GB || gpu.model == NvidiaGpu::A100_80GB {
+                warn!("MIG mode is Enabled in an Ampere GPU. Disabling Fabric Manager ...");
+                process::exit(1);
+            }
+        }
+    }
+
+    Ok(())
+}
+
+fn process_mig_config<T>(
+    model_str: &str,
+    gpu_indices: Vec<usize>,
+    mig_settings: &NvidiaMigConfig,
+) -> Result<()>
+where
+    T: MigGpuProfile,
+{
+    if gpu_indices.len() > 0 {
+        let default_profile = MigProfile::try_from("1").unwrap();
+        let gpu_model = NvidiaGpuModel::try_from(model_str).unwrap();
+        let mig_profile = mig_settings
+            .profile
+            .get(&gpu_model)
+            .unwrap_or(&default_profile);
+
+        info!("MIG Profile or the number of GPU slices: {:?}", mig_profile);
+        let json_string = serde_json::to_string(mig_profile.as_ref()).unwrap();
+        let profile =
+            serde_json::from_str::<T>(&json_string).context(error::JsonDeserializationSnafu)?;
+
+        set_mig_profile(gpu_indices, profile.get_mig_profile())?;
+    }
+
+    Ok(())
+}
+
+fn handle_mig_manager(mig_settings: NvidiaMigConfig, gpu_info: Vec<MigGpu>) -> Result<()> {
+    if mig_settings.device_partitioning_strategy == NvidiaDevicePartitioningStrategy::MIG {
```

Comment:
Rather than have both "enable" and "disable" paths in the same function, I would split this into two functions:
```
fn handle_mig_manager(mig_settings: NvidiaMigConfig, gpu_info: Vec<MigGpu>) -> Result<()> {
    if mig_settings.device_partitioning_strategy == NvidiaDevicePartitioningStrategy::MIG {
        enable_mig(mig_settings, gpu_info)
    } else {
        disable_mig(mig_settings, gpu_info)
    }
}
```